library/src/scala/main.scala

```diff
@@ -8,6 +8,192 @@
+    def doc: String
+
+    def usage: String
+  }
+  protected class SimpleArgument(val name: String, val typeName: String, val doc: String) extends Argument:
+    override def usage: String = s"[--$name] <$name>"
+  protected class OptionalArgument[T](val name: String, val typeName: String, val doc: String, val defaultValue: T)
+    extends Argument:
+    override def usage: String = s"[[--$name] <$name>]"
+  protected class VarArgument(val name: String, val typeName: String, val doc: String) extends Argument:
+    override def usage: String = s"[<$name> [<$name> [...]]]"
+
+  override type ArgumentParser[T] = util.CommandLineParser.FromString[T]
+  override type MainResultType = Any
+
+  /** Prints the main function's usage */
+  def usage(commandName: String, args: Seq[Argument]): Unit =
+    def wrappedArgumentUsages(argsUsage: List[String], maxLength: Int): List[String] =
+      def recurse(args: List[String], currentLine: String, acc: Vector[String]): Vector[String] =
+        (args, currentLine) match {
+          case (Nil, "") => acc
+          case (Nil, l) => (acc :+ l)
+          case (arg :: t, "") => recurse(t, arg, acc)
+          case (arg :: t, l) if l.length + 1 + arg.length <= maxLength => recurse(t, s"$l $arg", acc)
+          case (arg :: t, l) => recurse(t, arg, acc :+ l)
+        }
+
+      recurse(argsUsage, "", Vector()).toList
+    end wrappedArgumentUsages
+
+    val maxLineLength = 120 // TODO as parameter? As global Dotty parameter?
+
+    val usageBeginning = s"Usage: $commandName "
+    val argsOffset = usageBeginning.length
+    val argUsages = wrappedArgumentUsages(args.map(_.usage).toList, maxLineLength - argsOffset)
+
+    println(usageBeginning + argUsages.mkString("\n" + " " * argsOffset))
+
+  /** Prints an explanation about the function */
+  def explain(commandName: String, commandDoc: String, args: Seq[Argument]): Unit =
+    if (commandDoc.nonEmpty)
+      println(commandDoc)
+    if (args.nonEmpty) {
+      val argNameShift = 2
+      val argDocShift = argNameShift + 2
+
+      println("Arguments:")
+      for (arg <- args)
+        val argDoc = StringBuilder(" " * argNameShift)
+        argDoc.append(s"${arg.name} - ${arg.typeName}")
+
+        arg match {
+          case o: OptionalArgument[?] => argDoc.append(" (optional)")
+          case _ =>
+        }
+
+        if (arg.doc.nonEmpty) {
+          val shiftedDoc = arg.doc.split("\n").map(" " * argDocShift + _).mkString("\n")
+          argDoc.append("\n").append(shiftedDoc)
+        }
+
+        println(argDoc)
+    }
+
+  /** Runs the command and handles its return value */
+  def run(f: => MainResultType): Unit =
+    f match
+      case ExitCode(n) => System.exit(n)
+      case _ =>
+
+  override def command(args: Array[String], commandName: String, docComment: String): Command =
+    new Command(commandName, docComment):
+      /** A buffer of demanded arguments */
+      private var argInfos = new mutable.ListBuffer[self.Argument]
```

Comment:
It might be simpler to have and not introduce `Argument`
```suggestion
      private var argNames = new mutable.ListBuffer[String]
      private var argTypes = new mutable.ListBuffer[String]
      private var argKind = new mutable.ListBuffer[Int] // normal=0, default=1,  varargs=2
```

Reply:
Or 
```suggestion
      private var argNames = new mutable.ListBuffer[String]
      private var argTypes = new mutable.ListBuffer[String]
      private var varAgrs = false // is there a varargs arguments? We know this one is the last argument
      private var argsWithDefaults = 0 // number of arguments with defaults, we know they are the last ones before the vararg argument if any
```

Reply:
But with this, we would lose the ability to pattern match on arguments in `usage` and `explain`, rendering the help message generation more complex and harder for the user

Reply:
This is a change that would be worth if we make the `main` final. Then `usage` and `explain` would be private.